src/foam/nanos/http/AuthWebAgent.java

```diff
@@ -201,7 +213,11 @@ public void execute(X x) {
 
     if ( session != null && session.getContext() != null && session.getContext().get("user") != null ) {
       if ( auth.check(session.getContext(), permission_) ) {
-        getDelegate().execute(x.put(Session.class, session).put("user", session.getContext().get("user")));
+        if ( session.getContext().get("entity") != null ) {
```

Comment:
@adamvy This particular point has been debated back and forth, eventually deciding on X.user for authentication and X.agent for the agent the user is acting on behalf of. 
@deejes Please make use of Adam's simplified context setup.

Reply:
Or to put it different.  I can act as an account that is disabled, but in doing so, a disabled account probably won't be able to do anything.  The disabled state is not a factor on whether I can act as them or not.

Reply:
@jlhughes Sounds good, those names are reasonable.  Simplified context setup is the more important goal, rather than complex conditionals.

Reply:
> I think it would be more consistent if we always had x.user as the actual user who made the request, and x.effectiveUser as the agent/user that they are acting as.

Two things:

> and x.effectiveUser as the agent/user that they are acting as

The **entity** is the User being acted as. The **agent** is the User who is acting as the entity.

In the context, if a user is being acted as, then `x.get("user")` is the entity and `x.get("agent")` is the agent. If a user is not being acted as, then `x.get("user")` is just the user and `x.get("agent")` is null.

> x.user as the actual user who made the request

Who are you referring to by "actual user" here? The agent or the entity, using the definitions above?

In Ablii, we want to be doing everything as the Business, which is the entity, which is `x.get("user")`. That's because the id is used in many places such as relationships and authentication code. We want you to see the Business's invoices, contacts, and otherwise, not the user's.

I'm not sure how that translates to using the API, but I wanted to clarify our intentions when we designed it.

> IF we dont' want to go update all that code, we could add X.realUser as the actual user who made the request, and treat X.user as the effective user they are acting as.

We didn't want to update all that code, and that's exactly what we did. Just replace what you call `X.realUser` with `X.agent`. `X.user` is the business, the user you're acting as, and `X.agent` is you, which is rarely accessed.

Reply:
I appreciate the explanation, I should clarify that when I say that I find the code "confusing" it doesn't necessarily mean that am I actively "confused" but rather that I think that the code will likely lead to confusion of the next developer who comes along in 3 months and tries to understand it :).

> The entity is the User being acted as. The agent is the User who is acting as the entity.

My issue with that naming is that "entity" is a very general term, and we now have 3 names to describe 2 things.  We have "entity" "agent" and "user" but we only really use "agent" and "user"

My main objection/feedback on the PR is not the specific names we use (although I think there are clearer options) but that the code in this PR to construct the context uses a lot of contidionals.  Currently the code sets "entity" in a context in one place, then later pulls it out and uses it as a conditional on whether to put "agent" and "user" or just "user" in the context is unnecessary complication.  It would be safer and simpler to just always put "user" and "agent" in the context, then we never need to refer to "entity" in the context, which is a far too general name for this use case.

Reply:
I agree that the naming is confusing. Agent, entity, user all seem synonymous. Something along the lines of realUser and effectiveUser, and then just always populating both values would be easier to understand.